app/product_commands.go

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-server/v6/app/request"
+	"github.com/mattermost/mattermost-server/v6/model"
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+)
+
+// Code inspired on plugin_commands.go.
+//
+// Key differences/points:
+// - There's no need of health checks or unregisterProductCommands on products, they are compiled and assumed as active server side
+// - HooksForProduct still returns a plugin.Hooks struct, it might make sense to improve the name/package
+// - Plugin code had a check for a plugin crash after a command was executed, that has been omitted for products
+
+type ProductCommand struct {
+	Command   *model.Command
+	ProductID string
+}
+
+func (a *App) RegisterProductCommand(ProductID string, command *model.Command) error {
+	if command.Trigger == "" {
+		return errors.New("invalid command")
+	}
+	if command.AutocompleteData != nil {
+		if err := command.AutocompleteData.IsValid(); err != nil {
+			return errors.Wrap(err, "invalid autocomplete data in command")
+		}
+	}
+
+	// TODO: check this block
+	if command.AutocompleteData == nil {
+		command.AutocompleteData = model.NewAutocompleteData(command.Trigger, command.AutoCompleteHint, command.AutoCompleteDesc)
+	} else {
+		baseURL, err := url.Parse("/plugins/" + ProductID)
+		if err != nil {
+			return errors.Wrapf(err, "Can't parse url %s", "/plugins/"+ProductID)
+		}
+		err = command.AutocompleteData.UpdateRelativeURLsForPluginCommands(baseURL)
+		if err != nil {
+			return errors.Wrap(err, "Can't update relative urls for plugin commands")
+		}
+	}
+
+	command = &model.Command{
+		Trigger:              strings.ToLower(command.Trigger),
+		TeamId:               command.TeamId,
+		AutoComplete:         command.AutoComplete,
+		AutoCompleteDesc:     command.AutoCompleteDesc,
+		AutoCompleteHint:     command.AutoCompleteHint,
+		DisplayName:          command.DisplayName,
+		AutocompleteData:     command.AutocompleteData,
+		AutocompleteIconData: command.AutocompleteIconData,
+	}
+
+	a.ch.productCommandsLock.Lock()
+	defer a.ch.productCommandsLock.Unlock()
+
+	for _, pc := range a.ch.productCommands {
+		if pc.Command.Trigger == command.Trigger && pc.Command.TeamId == command.TeamId {
+			if pc.ProductID == ProductID {
+				pc.Command = command
+				return nil
+			}
+		}
+	}
+
+	a.ch.productCommands = append(a.ch.productCommands, &ProductCommand{
+		Command:   command,
+		ProductID: ProductID,
+	})
+	return nil
+}
+
+func (a *App) UnregisterProductCommand(productID, teamID, trigger string) {
```

Comment:
@isacikgoz I can't find the usage for this (and UnregisterPluginCommand) since:
- for plugins they are all unregistered when the plugin is disabled/uninstalled (not one by one)
- for products is done at build time, so unregisterCommand doesn't seem to be necessary

is that correct?

Reply:
Yes that seems to be correct to me, I don't see any use of unregistering a command.